### PR TITLE
Feature/sc 108358/implement add card to folder

### PR DESCRIPTION
**Pull request by markhornak**
Shortcut : [108358](https://app.shortcut.com/guru/story/108358/implement-add-card-to-folder)

### Overview

This ticket adds the ability add and move a card to a folder.  

move card requires a Card Id/Object, source folder id/object, and a target folder id/object
add card requires a Card id/object, target folder id/object

helper methods for move_card() and add_card() were also added to the Folder object.

### CODE REVIEW

File: guru/core.py [hunk 1]
```diff
@@ -2446,6 +2447,103 @@ def add_folder(self, title, collection, parentFolder=None, description=None):
...
+    # check if we have an object...
+    if not source_card_obj:
+      raise ValueError(f"couldn't find card in source_folder!: {card}")
+    # get the item_id from the source card object...
+    source_card_item_id = source_card_obj.item_id
+
+    # get folder object if possible...
+    target_folder_obj = self.get_folder(target_folder)
+    if not target_folder_obj:
+      raise ValueError(f"couldn't find target folder! : {target_folder}")
```
Comment (jahlbornG) [inline guru/core.py:2482]: nit, one thing i noticed is the old code uses logging with colors for errors, new code is throwing exceptions.  is throwing exceptions the plan moving forward?

File: guru/core.py [hunk 2]
```diff
@@ -2446,6 +2447,103 @@ def add_folder(self, title, collection, parentFolder=None, description=None):
...
+                "entryType": "card",
+                "id": source_card_item_id
+            }
+        ],
+        "prevSiblingItemId": "first"
+    }
+
+    # clear the cache for the folder since we moved a card...
+    self.__clear_cache(f"{self.base_url}/folders/{target_folder_slug}/items")
+
```
Comment (jahlbornG) [inline guru/core.py:2500]: do we need to modify the local items for the source/target folders if they are loaded? (same for add method)

File: guru/core.py [hunk 3]
```diff
@@ -2446,6 +2447,103 @@ def add_folder(self, title, collection, parentFolder=None, description=None):
     if status_to_bool(response.status_code):
       return Folder(response.json(), guru=self)
 
+  def move_card_to_folder(self, card, source_folder, target_folder):
+    """
+    Moves an existing in the collection card to another folder. The card will be added to the end
```
[the remark this answers is not in the document]
Comment (markhornak) [reply guru/core.py:2452]: bad copy/paste effort there...

File: guru/core.py (hunk 2 above)
Comment (markhornak) [reply guru/core.py:2500]: Added a .refresh() to the Folders object that clears out the items and calls the internal __get_items(refresh=True).  The refresh=True forces the __get_items to call get_folder_items() to rebuild the internal tuples off the new object.

Comment (jahlbornG) [reply guru/core.py:2500]: i'm curious why you went with another api call instead of just modifying the lists locally using the data at hand?

File: guru/data_objects.py [hunk 4]
```diff
@@ -97,7 +97,10 @@ def __init__(self, data, folder_items=[], guru=None, parent_folder=None):
     else:
       self.collection = None
 
-  # arrays to hold contents of the Folder, returned with
+  # arrays to hold contents of the Folder.  Note: items array is exposed for other methods to add objects
+  # such as cards or folders to the Folders object. e.g. add_card_to_folder() updates the items array
+  # and will do a save_folder() call.
+    self.items = []
     self.__all_items = []
```
[the remark this answers is not in the document]
Comment (markhornak) [reply guru/data_objects.py:104]: ok, **__all_items** is something in Boards that flattened the Sections out. Example, if there was a board with 2 sections and each section had 2 cards in them, the **items** list would have 2 items (each section), the **__all_items** list would have 4 (2 sections, 2 cards).

File: guru/core.py (hunk 2 above)
Comment (markhornak) [reply guru/core.py:2500]: I was/am having issues with name space munging of the __<lists> from inside the core.py.

File: guru/core.py [hunk 5]
```diff
@@ -2446,6 +2450,117 @@ def add_folder(self, title, collection, parentFolder=None, description=None):
...
+    self.__clear_cache(
+        f"{self.base_url}/folders/{clean_slug(source_folder_obj.slug)}/items")
+
+    url = f"{self.base_url}/folders/{target_folder_slug}/action"
+    response = self.__post(url, data)
+    if status_to_bool(response.status_code):
+      # refresh the internal items for the source Folder
+      source_folder_obj.update_lists(source_card_obj, "remove")
+      # return the response
+      return status_to_bool(response.status_code)
```
Comment (jahlbornG) [inline guru/core.py:2516]: this should not be nested under the if block right?  we want to return the status regardless of the result.

File: guru/core.py [hunk 6]
```diff
@@ -2446,6 +2450,117 @@ def add_folder(self, title, collection, parentFolder=None, description=None):
...
+
+    # clear the cache for the folder since we added a card...
+    self.__clear_cache(f"{self.base_url}/folders/{target_folder_slug}/items")
+
+    url = f"{self.base_url}/folders/{target_folder_slug}/action"
+    response = self.__post(url, data)
+    if status_to_bool(response.status_code):
+      # refresh the internal items for the source Folder
+      target_folder_obj.update_lists(card_obj, "add")
+      return status_to_bool(response.status_code)
```
Comment (jahlbornG) [inline guru/core.py:2562]: same comment as above, this shouldn't be nested under the if block

File: guru/data_objects.py [hunk 7]
```diff
@@ -144,15 +146,38 @@ def cards(self):
...
+    """
+    if action == 'remove':
+      if isinstance(obj, Card):
+        self.__cards.remove(obj)
+      elif isinstance(obj, Folder):
+        self.__folders.remove(obj)
+      else:
+        return
+      self.items.remove(obj)
+    elif action == "add":
```
Comment (jahlbornG) [inline guru/data_objects.py:167]: nit, inconsistent double vs single quotes for the constants.

File: guru/data_objects.py [hunk 8]
```diff
@@ -144,15 +146,38 @@ def cards(self):
...
+      if isinstance(obj, Card):
+        self.__cards.append(obj)
+      elif isinstance(obj, Folder):
+        self.__folders.append(obj)
+      else:
+        return
+      self.items.append(obj)
```
Comment (jahlbornG) [inline guru/data_objects.py:174]: i believe these should all be added first, not last.

File: guru/data_objects.py [hunk 9]
```diff
@@ -144,15 +146,38 @@ def cards(self):
...
 
   def __get_items(self):
     """
       method to load items for a Folder.  Useful if the intent is to keep the references to the Folders and sub-Folders in tact.  Loads items on a Folder for those folders that were not already retrieved with a get_folder(<slug>) call.
+
+      Args: refresh (Boolean, optional) - provides a means to force a refresh of folder items when things change like with a remove, add or move of objects w/in the folder_items
```
Comment (jahlbornG) [inline guru/data_objects.py:180]: this doesn't seem to correspond to the actual code

Comment (markhornak) [reply guru/data_objects.py:180]: Yup, left over from previous work, removed!

File: guru/core.py (hunk 6 above)
Comment (markhornak) [reply guru/core.py:2562]: Yes, moved out of the If block

File: guru/core.py (hunk 5 above)
Comment (markhornak) [reply guru/core.py:2516]: Yes, moved out of the If block

File: guru/core.py [hunk 10]
```diff
@@ -2446,6 +2450,117 @@ def add_folder(self, title, collection, parentFolder=None, description=None):
...
+    # clear the cache for the folder since we moved a card...
+    self.__clear_cache(f"{self.base_url}/folders/{target_folder_slug}/items")
+    self.__clear_cache(
+        f"{self.base_url}/folders/{clean_slug(source_folder_obj.slug)}/items")
+
+    url = f"{self.base_url}/folders/{target_folder_slug}/action"
+    response = self.__post(url, data)
+    if status_to_bool(response.status_code):
+      # refresh the internal items for the source Folder
+      source_folder_obj.update_lists(source_card_obj, "remove")
```
[the remark this answers is not in the document]
Comment (markhornak) [reply guru/core.py:2514]: Yeah, this is possible, it's a weird timing thing with my testing, where I didn't see the case where the target source didn't have the card b/c it was running the get again b/c cache was cleared for the folders/items.  But to be safe, I'll add the update_lists with the "add" for source_card_obj.

File: guru/data_objects.py (hunk 8 above)
Comment (jahlbornG) [reply guru/data_objects.py:174]: i think you misunderstood my comment.  i meant, the object that is added to the lists should be added to the beginning of the lists, not the end (i.e. prepend not append).  because the cards are added to the beginning of the folder.

Comment (markhornak) [reply guru/data_objects.py]: ah, gotcha, made it an .insert(0,obj) instead of .append.

File: guru/data_objects.py [hunk 11]
```diff
@@ -145,14 +147,40 @@ def cards(self):
     return tuple(self.__cards)
 
   @property
-  def items(self):
-    if not self.__has_items:
-      self.__get_items()
-    return tuple(self.__all_items)
```
Comment (jahlbornG) [inline guru/data_objects.py:151]: do you still want this change?  doesn't this expose the `items` property without the lazy loading behavior?  don't you want to revert this change an the one above which makes `__all_items` -> `items`?

File: guru/data_objects.py (hunk 8 above)
Comment (markhornak) [reply guru/data_objects.py]: fixed here: [11a6528](https://github.com/guruhq/py-sdk/pull/56/commits/11a65282411e638f344e27f49ee708fbdc24920d)

File: guru/data_objects.py (hunk 11 above)
Comment (markhornak) [reply guru/data_objects.py]: yeah, my bad, deleted the wrong method.  I had inserted an "has_items" method to track down the issue with the update_lists.  ugggh.

FIxed here: [0c0c7d2](https://github.com/guruhq/py-sdk/pull/56/commits/0c0c7d2025cd402624eed35210d7b660a2fcbf0c)

File: guru/data_objects.py (hunk 8 above)
Comment (markhornak) [reply guru/data_objects.py]: fixed here: [0c0c7d2](https://github.com/guruhq/py-sdk/pull/56/commits/0c0c7d2025cd402624eed35210d7b660a2fcbf0c)

**Review by jahlbornG — APPROVED** [review]
